File: cli/tauri.js/helpers/app-paths.js
const { existsSync } = require('fs')
const { resolve, join, normalize, sep } = require('path')

/**
 *
 * @returns {{length}|*}
 */
function getAppDir () {
  let dir = process.cwd()
  let count = 0

  // only go up three folders max
  while (dir.length && dir[dir.length - 1] !== sep && count <= 2) {
    if (existsSync(join(dir, 'tauri.conf.js'))) {
      return dir
    }
    count++
    dir = normalize(join(dir, '..'))
  }

  // just return the current directory
  return process.cwd()
}

const appDir = getAppDir()
const tauriDir = resolve(appDir, 'src-tauri')

module.exports = {
  appDir,
  tauriDir,

  resolve: {
    app: dir => resolve(appDir, dir),
    tauri: dir => resolve(tauriDir, dir)
  }
}

File: cli/src/helpers/app_paths.rs
use std::env::current_dir;
use std::path::PathBuf;

pub fn app_dir() -> PathBuf {
  let mut dir = current_dir().expect("failed to read cwd");

  let mut count = 0;

  // only go up three folders max
  while count <= 2 {
    let test_path = dir.join("src-tauri/tauri.conf.json");
    if test_path.exists() {
      return dir;
    }
    count += 1;
    match dir.parent() {
      Some(parent) => {
        dir = parent.to_path_buf();
      }
      None => break,
    }
  }

  panic!("Couldn't recognize the current folder as a Tauri project.")
}

pub fn tauri_dir() -> PathBuf {
  app_dir().join("src-tauri")
}
